### PR TITLE
Ensure SystemTextJsonHelper always HTML encodes output.

### PR DESCRIPTION
**Pull request by pranavkm**
Ensure JsonSerializer always HTML encodes output.

* Update JsonOptions.JsonSerializerOptions to use encoder scheme that does not encode non-ASCII
  characters by default. This makes the encoding comparable to Json.NET's defaults
* In SystemTextJsonHelper, ensure that the content is always HTML-encoded
* Unskip skipped test

Fixes https://github.com/aspnet/AspNetCore/issues/9946
Fixes https://github.com/aspnet/AspNetCore/issues/11459

------------------------------------------------------------------------------

## Description

System.Text.Json introduced a feature that allows configuring the encoding behavior including preventing encoding HTML content. In one case, MVC needs to guarantee the generated content is always HTML safe. This is important to prevent XSS.

## Regression

The feature was added to System.Text.Json fairly recently and this is a reaction.

## Customer Impact
n/a

## Risk
n/a

### CODE REVIEW

File: src/Mvc/Mvc.ViewFeatures/src/Rendering/SystemTextJsonHelper.cs [hunk 1]
```diff
@@ -10,20 +11,47 @@ namespace Microsoft.AspNetCore.Mvc.Rendering
...
             // JsonSerializer always encodes non-ASCII chars, so we do not need
             // to do anything special with the SerializerOptions
-            var json = JsonSerializer.Serialize(value, _options.JsonSerializerOptions);
+            var json = JsonSerializer.Serialize(value, _htmlSafeJsonSerializerOptions);
             return new HtmlString(json);
         }
+
+        private static JsonSerializerOptions CreateHtmlSafeSerializerOptions(JsonSerializerOptions serializerOptions)
+        {
+            var htmlSafeOptions = new JsonSerializerOptions
```
Comment (rynowak) [inline src/Mvc/Mvc.ViewFeatures/src/Rendering/SystemTextJsonHelper.cs]: @ahsonkhan @steveharter  - these are the kinds of cases where a copy constructor would be massively useful.

**Review by rynowak — APPROVED** [review]

File: src/Mvc/Mvc.Core/src/JsonOptions.cs [hunk 2]
```diff
@@ -26,6 +27,9 @@ public class JsonOptions
 
             // Use camel casing for properties
             PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
+
+            // Does not encode all non-ASCII characters which matches Json.NET"s behavior.
+            Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
```
Comment (GrabYourPitchforks) [inline src/Mvc/Mvc.Core/src/JsonOptions.cs]: I believe this is a dangerous place to put this line of code. The _UnsafeRelaxedJsonEscaping_ property is __only__ intended to be used in specific scenarios where the stack has tight control over how the data will be written to the wire and how the client will consume the data. Any other usage is dangerous, hence the term "unsafe" in the name.

Here, the part that causes me concern is that ASP.NET is configuring this setting in a place very far removed from its intended known-good usage scenario. Specifically, this particular escaping should only be set if we know for absolute certain that the response will contain a header `Content-Type: application/json; charset=utf=8`, but this line of code is so far removed from the component that sets the response Content-Type header that this risks introducing a vulnerability due to misuse.

Comment (blowdart) [reply src/Mvc/Mvc.Core/src/JsonOptions.cs]: What the heck is the point in having safe defaults if asp.net is going to dump all over them and turn it off?

Comment (pranavkm) [reply src/Mvc/Mvc.Core/src/JsonOptions.cs]: We might be able to localize this to the Json Formatter when the encoding is Uf8 and leave the defaults untouched. Would that be sufficient?

File: src/Mvc/Mvc.ViewFeatures/src/Rendering/SystemTextJsonHelper.cs (hunk 1 above)
Comment (JamesNK) [reply src/Mvc/Mvc.ViewFeatures/src/Rendering/SystemTextJsonHelper.cs]: +1 as a future feature.

File: src/Mvc/Mvc.Core/src/Formatters/SystemTextJsonOutputFormatter.cs [hunk 3]
```diff
@@ -33,6 +33,19 @@ public SystemTextJsonOutputFormatter(JsonOptions options)
             SupportedMediaTypes.Add(MediaTypeHeaderValues.ApplicationAnyJsonSyntax);
         }
 
+        internal static SystemTextJsonOutputFormatter CreateFormatter(JsonOptions jsonOptions)
+        {
+            var jsonSerializerOptions = jsonOptions.JsonSerializerOptions;
+
+            if (jsonSerializerOptions.Encoder is null)
```
Comment (pranavkm) [inline src/Mvc/Mvc.Core/src/Formatters/SystemTextJsonOutputFormatter.cs]: 🆙 📅 

This limits the encoder to use the relaxed encoding specifically in the output formatter assuming the user hasn't already configured it.

File: src/Mvc/Mvc.ViewFeatures/src/Rendering/SystemTextJsonHelper.cs (hunk 1 above)
Comment (ahsonkhan) [reply src/Mvc/Mvc.ViewFeatures/src/Rendering/SystemTextJsonHelper.cs]: I see. Being tracked by an issue that @pranavkm opened earlier: https://github.com/dotnet/corefx/issues/39939

File: src/Mvc/Mvc.Core/src/Formatters/SystemTextJsonOutputFormatter.cs (hunk 3 above)
Comment (pranavkm) [reply src/Mvc/Mvc.Core/src/Formatters/SystemTextJsonOutputFormatter.cs]: Spoke to @GrabYourPitchforks offline. We decided limited this to the formatter is reasonable. The JsonHelper always defaults to using the Default encoding.

File: src/Mvc/Mvc.Core/src/Infrastructure/JsonSerializerOptionsCopyConstructor.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+                IgnoreNullValues = serializerOptions.IgnoreNullValues,
+                IgnoreReadOnlyProperties = serializerOptions.IgnoreReadOnlyProperties,
+                MaxDepth = serializerOptions.MaxDepth,
+                PropertyNameCaseInsensitive = serializerOptions.PropertyNameCaseInsensitive,
+                PropertyNamingPolicy = serializerOptions.PropertyNamingPolicy,
+                ReadCommentHandling = serializerOptions.ReadCommentHandling,
+                WriteIndented = serializerOptions.WriteIndented
+            };
+
+            for (var i = 0; i < serializerOptions.Converters.Count; i++)
```
Comment (DAllanCarr) [inline src/Mvc/Mvc.Core/src/Infrastructure/JsonSerializerOptionsCopyConstructor.cs]:
```suggestion
            for (var converterIdx = 0; converterIdx < serializerOptions.Converters.Count; converterIdx++)
```

Comment (pranavkm) [reply src/Mvc/Mvc.Core/src/Infrastructure/JsonSerializerOptionsCopyConstructor.cs]: I'm not sure this makes the code any better.

**Pull request merged**